partners_api/booking_api.cpp

```diff
@@ -45,229 +234,128 @@ void BookingApi::GetMinPrice(string const & hotelId, string const & currency,
+    if (pos != string::npos)
+      langCode = lang.substr(0, pos);
+    else
+      langCode = lang;
+  }
+
+  ss << "/" << hotelId << "/" << langCode << ".json";
+
+  return RunSimpleHttpRequest(ss.str(), result);
+}
+
+string Api::GetBookHotelUrl(string const & baseUrl) const
+{
+  return GetDescriptionUrl(baseUrl) + "#availability";
+}
+
+string Api::GetDescriptionUrl(string const & baseUrl) const
+{
+  return baseUrl + string("?aid=") + BOOKING_AFFILIATE_ID;
+}
+
+void Api::GetMinPrice(string const & hotelId, string const & currency,
+                      GetMinPriceCallback const & fn)
+{
+  m_lastHotelId = hotelId;
+
+  auto const callback = [this, fn](string const & hotelId, string const & price,
+                                   string const & currency) {
+    GetPlatform().RunOnGuiThread([this, hotelId, price, currency, fn]() {
+      if (m_lastHotelId != hotelId)
+        return;
+
+      fn(price, currency);
+    });
+  };
+
+  auto const testingMode = m_testingMode;
 
+  threads::SimpleThread([hotelId, currency, callback, testingMode]()
+  {
     string minPrice;
     string priceCurrency;
+    string httpResult;
+    if (!RawApi::GetHotelAvailability(hotelId, currency, httpResult, testingMode))
+    {
+      callback(hotelId, minPrice, priceCurrency);
+      return;
+    }
+
     try
     {
-      my::Json root(answer.Data().c_str());
-      if (!json_is_array(root.get()))
-        MYTHROW(my::Json::Exception, ("The answer must contain a json array."));
-      size_t const sz = json_array_size(root.get());
-
-      if (sz > 0)
-      {
-        // Read default hotel price and currency.
-        auto obj = json_array_get(root.get(), 0);
-        my::FromJSONObject(obj, "min_price", minPrice);
-        my::FromJSONObject(obj, "currency_code", priceCurrency);
-
-        // Try to get price in requested currency.
-        if (!currency.empty() && priceCurrency != currency)
-        {
-          json_t * arr = json_object_get(obj, "other_currency");
-          if (arr && json_is_array(arr))
-          {
-            size_t sz = json_array_size(arr);
-            for (size_t i = 0; i < sz; ++i)
-            {
-              auto el = json_array_get(arr, i);
-              string code;
-              my::FromJSONObject(el, "currency_code", code);
-              if (code == currency)
-              {
-                priceCurrency = code;
-                my::FromJSONObject(el, "min_price", minPrice);
-                break;
-              }
-            }
-          }
-        }
-      }
+      FillPriceAndCurrency(httpResult, currency, minPrice, priceCurrency);
     }
     catch (my::Json::Exception const & e)
     {
       LOG(LERROR, (e.Msg()));
       minPrice.clear();
       priceCurrency.clear();
     }
-    fn(minPrice, priceCurrency);
-    m_request.reset();
-  };
-
-  m_request.reset(downloader::HttpRequest::Get(url, callback));
+    callback(hotelId, minPrice, priceCurrency);
+  }).detach();
```

Comment:
Why not to create a single thead with simple synchronous queue to process get min price requests, instead of this uncontrolled threads allocation?